### PR TITLE
Possible race in creation / parenting of windows

### PR DESCRIPTION
**Pull request by tonikitoo**
fixup! c++ / mojo changes for 'external window mode'

Patch makes WindowTreeHostFactory an associated interface of WindowTree.
This allows both interfaces to share the same mojo pipe, ensuring ordering
of messages.

This is important because during the creation of windows/displays,
WindowTreeHostFactory::CreatePlatformWindow messages concurs with
subsequent WindowTree method calls, causing non-deterministic crashes
in chrome/mus.

In summary, the patch:

- renames WindowTreeHostFactoryRegistrar to ExternalWindowTreeFactory.
- removes the use/logic of the WindowTreeHostFactoryRequest from the
  ::Registrar method.
- adds a WindowTree::GetWindowTreeHostFactory method that takes an
  associated interface request.

Issue #161.

### CODE REVIEW

**Comment by tonikitoo** [comment]
PENDING: update unittests code.

File: services/ui/manifest.json [hunk 1]
```diff
@@ -66,7 +66,7 @@
           "ui::mojom::WindowTreeHostFactory"
         ],
         "window_tree_host_factory_registrar": [
```
Comment (msisov) [inline services/ui/manifest.json]: nit: s/window_tree_host_factory_registrar/external_window_tree_factory/"

File: services/ui/ws/display_binding.h [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include "base/macros.h"
 #include "mojo/public/cpp/bindings/binding.h"
 #include "services/ui/public/interfaces/window_tree_host.mojom.h"
+#include "services/ui/public/interfaces/window_tree.mojom.h"
```
Comment (msisov) [inline services/ui/ws/display_binding.h]: is this include really used here? Maybe better to add it to a .cc file of the ancestor? This would avoid unnecessary includes in other headers, which inherit this interface.

**Review by msisov — COMMENTED** [review]
I got how it works. Good solution. Waiting for unittests. Left a couple of nits.

File: services/ui/ws/display_binding.h (hunk 2 above)
Comment (tonikitoo) [reply services/ui/ws/display_binding.h]: good catch. it is not needed. removed.

File: services/ui/manifest.json (hunk 1 above)
Comment (tonikitoo) [reply services/ui/manifest.json]: done

**Comment by tonikitoo** [comment]
reuploaded a final version. unittests passing.

File: services/ui/public/interfaces/window_tree.mojom [hunk 3]
```diff
@@ -282,6 +283,9 @@ interface WindowTree {
   // Stacks the window above all sibling windows.
   StackAtTop(uint32 change_id, uint32 window_id);
 
+  // TODO
```
Comment (msisov) [inline services/ui/public/interfaces/window_tree.mojom]: TODO? :) Do you mean to add a comment here?

File: services/ui/ws/window_tree.h [hunk 4]
```diff
@@ -685,6 +690,11 @@ class WindowTree : public mojom::WindowTree,
   mojom::WindowManager* window_manager_internal_;
   std::unique_ptr<WindowManagerState> window_manager_state_;
 
+  // TODO
```
Comment (msisov) [inline services/ui/ws/window_tree.h]: same here. Do you mean to add a comment?

**Review by msisov — APPROVED** [review]
lgtm

File: services/ui/ws/window_tree.h (hunk 4 above)
Comment (tonikitoo) [reply services/ui/ws/window_tree.h]: done

File: services/ui/public/interfaces/window_tree.mojom (hunk 3 above)
Comment (tonikitoo) [reply services/ui/public/interfaces/window_tree.mojom]: done
